org.adempiere.base/src/org/compiere/model/MInOut.java

```diff
@@ -1523,6 +1524,58 @@ public String prepareIt()
 		if (m_processMsg != null)
 			return DocAction.STATUS_Invalid;
 
+		//Check if Order is Valid - load all C_Order_ID into HashSet and replace/remove C_Order_ID if not valid
+		if(getC_Order_ID() > 0) {			
+			
+			HashSet<Integer> temp_C_Order_ID = new HashSet<Integer>(); // Load C_Order_IDs into Hash Set
+			
+			for(MInOutLine line : getLines())	{
+				
+				if(line.getC_OrderLine_ID() == 0) {
+					temp_C_Order_ID.add(0); // Add zero C_Order_ID
+				}
+				else {
+					temp_C_Order_ID.add(line.getC_OrderLine().getC_Order_ID());
```

Comment:
The call to getC_OrderLine() is very expensive.

I think this code can be made a lot simpler like this:
```java
if (getC_Order_ID() > 0) {
    int[] orderIds = DB.getIDsEx(get_TrxName(), "SELECT DISTINCT C_Order_ID FROM C_OrderLine WHERE C_Order_ID=?", getC_Order_ID());
    if (orderIds.length == 1 && orderIds[0] != getC_Order_ID())
        setC_Order_ID(orderIds[0]);
    else if (orderIds.length > 1)
        setC_Order_ID(0);
}
```

Reply:
Hi @CarlosRuiz-globalqss - but in this case, wouldn't `SELECT DISTINCT C_Order_ID FROM C_OrderLine WHERE C_Order_ID=?` return always the C_Order_ID set on our M_InOut?

The point would be to check if M_InOutLine -> C_OrderLine -> C_Order_ID == M_InOut.C_Order_ID, I mean something like this:
`SELECT DISTINCT ol.C_Order_ID FROM M_InOutLine iol JOIN C_OrderLine ol ON (iol.C_OrderLine_ID=ol.C_OrderLine_ID) WHERE iol.M_InOut_ID=?`

Or am I missing something?

Reply:
Thanks @CarlosRuiz-globalqss, I've committed the changes. With this we removed 41 lines of code and the behaviour is the same :)